components/script/dom/servoparser/mod.rs

```diff
@@ -463,6 +483,35 @@ impl ServoParser {
     }
 
     fn push_bytes_input_chunk(&self, chunk: Vec<u8>) {
+        // BOM sniff. This is needed because NetworkDecoder will switch the
+        // encoding based on the BOM, but it won't change
+        // `self.document.encoding` in the process.
+        {
+            let set_bom_to_none = if let Some(partial_bom) = self.bom_sniff.borrow_mut().as_mut() {
+                if partial_bom.len() + chunk.len() >= 3 {
+                    partial_bom.extend(chunk.iter().take(3 - partial_bom.len()).copied());
+                    //println!("Full BOM: {:?}", partial_bom);
+                    if let Some((encoding, _)) = Encoding::for_bom(&partial_bom) {
+                        //println!("Encoding: {}", encoding.name());
+                        self.document.set_encoding(encoding);
+                    } else {
+                        //println!("Bytes are not a BOM.");
+                    };
+                    true
+                } else {
+                    partial_bom.extend(chunk.iter().copied());
+                    //println!("Partial BOM: {:?}", partial_bom);
+                    false
+                }
+            } else {
+                //println!("partial_bom is None");
+                false
+            };
+            if set_bom_to_none {
+                *self.bom_sniff.borrow_mut() = None;
+            }
```

Comment:
I just noticed there are some commented `println` which might be good to remove them.

Besides, as the `set_bom_to_none` only lives in this scope, maybe we can just set it in the `true` case? then, we don't need to return the boolean value. 👀

```suggestion
            let set_bom_to_none = if let Some(partial_bom) = self.bom_sniff.borrow_mut().as_mut() {
                if partial_bom.len() + chunk.len() >= 3 {
                    partial_bom.extend(chunk.iter().take(3 - partial_bom.len()).copied());

                    if let Some((encoding, _)) = Encoding::for_bom(&partial_bom) {
                        self.document.set_encoding(encoding);
                    }

                    *self.bom_sniff.borrow_mut() = None;
                } else {
                    partial_bom.extend(chunk.iter().copied());
                }
            } 
```

Reply:
I didn't realize I hadn't cleaned up this part of the code. I'll file a PR to clean it up.

Reply:
Now that I look into it, that would be borrowing `self.bom_sniff` inside the scope of the other borrow, which causes a panic. That's why I did that by setting a variable, so the borrow that replaced `self.bom_sniff` with `None` was outside the scope.

I'm not sure if there's a more idiomatic way to do something like this.

Reply:
> I didn't realize I hadn't cleaned up this part of the code. I'll file a PR to fix it.

Thank you!

> That's why I did that by setting a variable, so the borrow that replaced `self.bom_sniff` with `None` was outside the scope.

Ah, I see. Then maybe let's keep it as is 🙇